### PR TITLE
Support embedded formatting in template literals annotated with `as const`

### LINKED ISSUE #15315: /* GraphQL */ comment tag does not format with typescript `as const`
**Prettier 3.0.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAlA8gIVwBUBlAfQEUBVAUWwE0yB1ASSIAkyBhXAORKKYAvJgD0AKkwBxAE4BDAA4ALCgBlM40ZgAGAHSiZMARwCucGQE9MJTMExolEBZgC++7ZjlpMkKBgDc+vq+GDgExOTUdPTCYpKyiirqmjr6hqbmVjZ2Dk6u7oFQIAA0IE4wAJboyKByMjIQAO4ACnUIaMggcgA2jXIWHaUARvJgANZwMCRyALZwqhVQcMgAZj1ocMOjE1MKcmCLAObIMDJmpRszFSdnmyBoR91wFCYQ8KvrdwBWaAAeJI9nq93kg1t0NqVTG84M0GgoOigvABaJZwAAm6JKIFOcgq3SOXAgMxmck6PW6WIeUEOTwAgjBThUhiZ4M1zAslh9wXclDAZt0mEoKvA0HswHASO1hRUAG7CiydMBoQYgGVmFhQDGwEhgGQVBQwWmakgwCxPLkQkAKBobJjyBSda1wDYyGXLUqLF0wWFyQ4ki13PYyF2dIZyIZwbpImAq62LGBMCpomBKZAADgADKUZHBTBUcz6-aTQZ9SjBw4nk6mkAAmUomDZEcMIsGWuAzCNojFo1RyakmX1wABiEBkJIZRzJLIgIBcLiAA)

```sh
--parser babel-ts
```

**Input:**

```tsx
const ROBOTS_QUERY_WITH_CONST = /* GraphQL */ `
  query S { shop }
` as const;

const ROBOTS_QUERY = /* GraphQL */ `
  query S { shop }
`;

```

**Output:**

```tsx
const ROBOTS_QUERY_WITH_CONST = /* GraphQL */ `
  query S { shop }
` as const;

const ROBOTS_QUERY = /* GraphQL */ `
  query S {
    shop
  }
`;

```

**Expected behavior:**

Should get the same formated result within the tagged string.

```
const ROBOTS_QUERY_WITH_CONST = /* GraphQL */ `
  query S { shop }
` as const;

const ROBOTS_QUERY = /* GraphQL */ `
  query S { shop }
`;

```

Of course, when it doesn't work with `as const`, you can drop that `as const` because it does not affect typescript's typing. However, there are places where dropping such `as const` might affect devs' experience

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #15315

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/embed/utils.js [hunk 1]
```diff
@@ -66,8 +66,18 @@ function hasLanguageComment(node, languageName) {
   );
 }
 
+function isAsConstExpression(node) {
+  return (
+    node.type === "TSAsExpression" &&
+    node.typeAnnotation.type === "TSTypeReference" &&
+    node.typeAnnotation.typeName.type === "Identifier" &&
+    node.typeAnnotation.typeName.name === "const"
+  );
```
Comment (fisker) [inline src/language-js/embed/utils.js:75]: We already merge  https://github.com/prettier/prettier/pull/15130?notification_referrer_id=NT_kwDOAAKiKLE3MTA1ODk0MDM3OjE3MjU4NA#event-10389275756, so we probably should add support for Flow parser too.

File: src/language-js/embed/graphql.js [hunk 2]
```diff
@@ -108,6 +108,9 @@ function printGraphqlComments(lines) {
 function isGraphQL({ node, parent }) {
   return (
     hasLanguageComment(node, "GraphQL") ||
+    (parent &&
```
Comment (fisker) [inline src/language-js/embed/graphql.js:111]: I don't think it's possible to be root.

File: src/language-js/embed/graphql.js [hunk 3]
```diff
@@ -108,6 +108,9 @@ function printGraphqlComments(lines) {
...
+      isAsConstExpression(parent) &&
+      hasLanguageComment(parent, "GraphQL")) ||
```
Comment (fisker) [inline src/language-js/embed/graphql.js:113]: I think we can move parent check logic into  `hasLanguageComment` for easier use.

File: tests/format/flow/as-const/jsfmt.spec.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+run_spec(import.meta, ["flow"], {
+  errors: {
+    "babel-flow": ["embedded.js"],
```
Comment (sosukesuzuki) [inline tests/format/flow/as-const/jsfmt.spec.js:3]: Babel cannot parse `as const`, I'll report it later.

Comment (fisker) [reply tests/format/flow/as-const/jsfmt.spec.js:3]: Newly added flow syntax all not supportted. It's fine.

**Review by fisker — APPROVED** [review]

**Pull request merged**